### PR TITLE
#64 Sign & Send Transactions (Ledger Wallet)

### PR DESCRIPTION
**Pull request by lewis-sqa**
## This PR Includes

- Refactored `LedgerWallet`
  - Added integration with `LedgerClient`. This abstracts away the details of `@ledgerhq/hw-transport-webhid` and makes it somewhat easier to handle. Unit tests to be written in a follow-up task for the interest of getting this fix over the line.
  - Added `accountId` to `localStorage` to avoid issues with the `getAccount` method.
  - Added support for redirecting to NEAR Wallet if the selected public key is not attached to the account.
  - Added full support for signing transactions with multiple actions.
  - Modified the signing process to fetch the nonce each time instead of only once during initialisation and manually incrementing thereafter.
- Refactored the `<Modal />` component to streamline the Ledger flow, standardise the event handling and clean up the state management.
- Modified the [`finality`](https://docs.near.org/docs/api/rpc#using-finality-param) of the `callFunction` method inside `ProviderService`. This fixes a bug where calling `getMessages` after `sendMessage` would often result in no UI change.
- Fixed the remaining `any` usages and tightened up the ESLint to be an error going forward.
- Created a task to address the lack of unit tests around `LedgerClient`: #126

### CODE REVIEW

File: src/wallets/hardware/LedgerWallet.ts [hunk 1]
```diff
@@ -254,95 +200,81 @@ export default class LedgerWallet
...
         keyType: transaction.publicKey.keyType,
         data: signature,
       }),
     });
 
-    const res = await this.provider.sendTransaction(signedTransaction);
-
-    if (typeof res.status !== "string") {
-      const successValue = res.status.SuccessValue || "";
+    return this.provider.sendTransaction(signedTx!).then((res) => {
```
Comment (lewis-sqa) [inline src/wallets/hardware/LedgerWallet.ts:267]: Do we need the `!` here?

Comment (lewis-sqa) [reply src/wallets/hardware/LedgerWallet.ts:267]: Need to figure out how to correctly handle `sendTransaction` responses

File: src/wallets/hardware/LedgerClient.ts [hunk 2]
```diff
@@ -0,0 +1,153 @@
...
+}
+
+interface SignParams {
+  data: Uint8Array;
+  derivationPath: string;
+}
+
+interface EventMap {
+  // TODO: Determine event data.
+  disconnect: unknown;
```
Comment (lewis-sqa) [inline src/wallets/hardware/LedgerClient.ts:56]: Need to test this event to determine the interface

**Comment by lewis-sqa** [comment]
Found a couple of bugs after some more testing:
- State doesn't update if you disconnect from the device instead of the sign out button.
- If you logged in with ledger previously, the page will crash. I think I need to be "lazy" about connecting back to the Ledger. I'm thinking I'll need to prompt at the point of signing to avoid the error.

File: src/wallets/hardware/LedgerWallet.ts (hunk 1 above)
Comment (lewis-sqa) [reply src/wallets/hardware/LedgerWallet.ts]: > Need to figure out how to correctly handle `sendTransaction` responses

This really needs addressing across all the wallet implementations.

File: src/wallets/hardware/LedgerClient.ts [hunk 3]
```diff
@@ -0,0 +1,152 @@
...
+  listen(callback: (data: Log) => void) {
+    const unsubscribe = listen(callback);
+
+    return {
+      remove: () => unsubscribe(),
+    };
+  }
+
+  setScrambleKey(key: string) {
+    this.transport.setScrambleKey(key);
```
Comment (lewis-sqa) [inline src/wallets/hardware/LedgerClient.ts:78]: Well this method does a whole lot of nothing: https://github.com/LedgerHQ/ledgerjs/blob/master/packages/hw-transport-webhid/src/TransportWebHID.ts#L248

**Review by amirsaran3 — APPROVED** [review]

**Pull request merged**